### PR TITLE
Add highlighting for embedded JavaScript/CSS in SVG files

### PR DESCRIPTION
**Pull request by Alhadis**
I hope this isn't introducing too many drastic changes to the grammar (if it is, I'll know not to bother with my proposed fix for [this](https://github.com/atom/language-gfm/issues/70)...)

Since XML is used to serialise every fathomable sort of data, I took a conservative and cautious approach - specifically targeting the contents of `<svg></svg>` pairs, and even then, only targeting `<script>` tags without a foreign language identifier:

Figure 1

Which means the grammar plays it safe with any `<script>` tag not explicitly flagged as JavaScript:

Figure 2

The assertion for detecting JavaScript is written [using extended notation](https://github.com/atom/language-xml/pull/54/files#diff-981311489a407bc2bdafe52940f7cadeR81) in an attempt to facilitate code review.

### CODE REVIEW

**Comment by pchaigno** [comment]
Hi @Alhadis! Could you add some [tests](https://github.com/atom/language-xml/blob/master/spec/xml-spec.coffee) for your changes?

**Comment by Alhadis** [comment]
Goodness me, where are my manners...? Fixed.

**Comment by Alhadis** [comment]
Just like the entire React community, the sight of JavaScript made me completely forget CSS exists.

I'm in the process of amending the specs to test for correct SVG/`<style>` handling now, but I felt impelled to leave some notes while they're drifting around in my skull:

1. The rule for matching embedded `<script>` tags has been given its own named rule in the repository. I figure if anybody extending the XML grammar needs to include embedded JS highlighting inside other tags, we can at least make doing so easy for them.

2. I haven't given CSS the same "runaway check" as I added for JS. The HTML grammar doesn't check for weird constructions like this, so I don't believe we should either:
```css
body::after{
	content: "lol </style> wat"
}
```
Furthermore, this would prevent the entire SVG file from rendering correctly, so it's highly unlikely there's actually any SVG code out there with that sort of syntactic abomination swimming around in its source.

**Comment by 50Wliu** [comment]
Sorry for the delay in reviewing this.  I'll try to get to it as soon as I can.

**Comment by Alhadis** [comment]
> Sorry for the delay in reviewing this. I'll try to get to it as soon as I can.



So, uh...

**Comment by 50Wliu** [comment]
> So, uh...

This is a **very** large PR with far-reaching implications that I don't have the resources (or time) to review right now.

**Comment by Alhadis** [comment]
All good, mate, it just sounded like you were about to review it within the next few days, and I thought you might've forgotten... :) No rush.

File: grammars/xml.cson [hunk 1]
```diff
@@ -70,156 +70,188 @@
...
-        'name': 'variable.language.documentroot.xml'
-    'end': '\\s*(>)'
-    'name': 'meta.tag.sgml.doctype.xml'
-    'patterns': [
+    'include': '#generic-xml'
+  }
+]
+'repository':
+  'generic-xml':
+    patterns: [
```
Comment (50Wliu) [inline grammars/xml.cson]: quote it

File: grammars/xml.cson [hunk 2]
```diff
@@ -70,156 +70,188 @@
...
-    'begin': '(<\\?)\\s*([-_a-zA-Z0-9]+)'
-    'captures':
-      '1':
-        'name': 'punctuation.definition.tag.xml'
-      '2':
-        'name': 'entity.name.tag.xml'
-    'end': '(\\?>)'
-    'name': 'meta.tag.preprocessor.xml'
+    'name': 'meta.svg.xml'
+    'begin': '(?i:<svg(?=[\\s:>])[^>]*>)'
```
Comment (50Wliu) [inline grammars/xml.cson]: `<Svg>` is different from `<svg>` as XML is case-sensitive, right?

File: grammars/xml.cson [hunk 3]
```diff
@@ -343,3 +375,94 @@
...
+          [^>]*
+          >
+        """
+        'beginCaptures':
+          '0':
+            'patterns': [
+              'include': '#generic-xml'
+            ]
+        'contentName': 'source.css.embedded.xml'
+        'end': '(?ix:</style>)'
```
Comment (50Wliu) [inline grammars/xml.cson]: If we match case-insensitively here then the beginCapture should as well.
Also, the `x` modifier is complete overkill.

File: grammars/xml.cson [hunk 4]
```diff
@@ -343,3 +375,94 @@
...
+          [^>]*
+          >
+        """
+        'beginCaptures':
+          '0':
+            'patterns': [
+              'include': '#generic-xml'
+            ]
+        'contentName': 'source.js.embedded.xml'
+        'end': '(?ix:</script>)'
```
Comment (50Wliu) [inline grammars/xml.cson]: Ditto.

**Comment by 50Wliu** [comment]
Initial review comments:
1. I don't see why you had to extract everything into a `#generic-xml` pattern (see language-html).
2. The specs would be much easier to review if you didn't add the helper functions and kept it as-is.
3. Regarding the "runaway" strings, I have an improvement coming in first-mate that should help out with that.

**Comment by Alhadis** [comment]
> I don't see why you had to extract everything into a #generic-xml pattern (see language-html).

For control, and so the scopes for opening/closing SVG tags will always share those of other SVG tags. Note the way I've included `#generic-xml` in the opening/closing tags of the SVG pattern.

> Regarding the "runaway" strings, I have an improvement coming in first-mate that should help out with that.

Uh-oh. Uh, what kind of improvement? I remember I've exploited things like this in the past to work around TextMate limitations...

I suppose I'd better have a look.

File: grammars/xml.cson (hunk 2 above)
Comment (Alhadis) [inline grammars/xml.cson]: Right, good catch. Totally forgot about that.

File: grammars/xml.cson (hunk 3 above)
Comment (Alhadis) [inline grammars/xml.cson]: What do you mean? We *are* [matching case-insensitively](https://github.com/Alhadis/language-xml/blob/ff0b98fd5d90baaa07b0fad3ce472d039feca8fa/grammars/xml.cson#L381-L401).

Comment (50Wliu) [inline grammars/xml.cson]: whoops

**Comment by 50Wliu** [comment]
atom/first-mate#78, coupled with atom/language-javascript#438 (which I'll be updating shortly to include strings as well).

**Comment by Alhadis** [comment]
Ah, it's injections-specific. All good then.

So... is there anything left for me to do here?

**Comment by 50Wliu** [comment]
> So... is there anything left for me to do here?

Revert the specs to their original format.

**Comment by Alhadis** [comment]
... okay, then...

**Comment by Alhadis** [comment]
@50Wliu Erm, the changes were applied 10 days ago... is there anything left for me to do here?

**Comment by 50Wliu** [comment]
I'll tell you when I get back to it.

**Comment by Alhadis** [comment]
Closing for reasons explained by https://github.com/atom/language-javascript/pull/497#discussion_r114573368.

**Pull request closed without merging**